docs/propagation.md

```diff
@@ -0,0 +1,227 @@
+```
+
+Here both dimensions 0 and 1 of the input correspond to dimension 0 of the output. Say we start by giving factors to the input:
+
+```
+(i,j,k) : i=2, j=4, k=32
+```
+
+You can see that if we want to use the same factors for the output, we would need a single dimension to reference multiple factors:
+
+```
+(i,j,k) -> ((ij), k) : i=2, j=4, k=32
+```
+
+The same can be done if the reshape were to split a dimension:
+
+```
+%out = mhlo.reshape(%in) : (tensor<8x32xf32>) -> tensor<2x4x32xf32> ((ij), k) -> (i,j,k) : i=2, j=4, k=32
+```
+
+The dimension of size 8 here is essentially composed of the factors 2 and 4, which is why we are calling the factors (i,j,k) factors.
+
+These factors can also work with cases where there is no full dimension that corresponds to one of the factors:
+
+```
+%out = mhlo.reshape(%in) : (tensor<8x4xf32>) -> tensor<2x16xf32> ((ij), k) -> (i,(jk)) : i=2, j=4, k=4
+```
+
+This example also emphasizes why we need to store the factor sizes - since we can't easily deduce them from the corresponding dimensions.
+
+### Core Propagation Algorithm
+
+#### Propagate shardings along factors
+
+In Shardy, we have the hierarchy of tensors, dimensions, and factors. They represent data at different levels. A factor is a sub-dimension. It is an internal hierarchy used in sharding propagation. Each dimension may correspond to one or more factors. The mapping between dimension and factor is defined by OpShardingRule.
+
+![Schema showing the Shardy propagation algorithm.](images/propagation_algorithm.png)
+
+**Shardy propagates sharding axes along factors instead of dimensions**. To do that, we have three steps as shown in the figure below
+
+1. Project DimSharding to FactorSharding
+2. Propagate sharding axes in the space of FactorSharding
+3. Project the updated FactorSharding to get the updated DimSharding
+
+![Schema showing sharding propagation across FactorSharding and DimSharding.](images/projected_sharding.png)
+
+#### Visualization of Sharding Propagation Along Factors
+
+We will use the following table to visualize the sharding propagation problem and algorithm.
+
+|  | F0 | F1 | F2 | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- |
+| T0 |  |  |  |  |
+| T1 |  |  |  |  |
+| T2 |  |  |  |  |
+
+* Each column represents a factor. F0 means the factor with index 0. We propagate shardings along factors (columns).
+* Each row represents a tensor. T0 refers to the tensor with index 0. Tensors are all operands and results involved for a specific operation. The axes in a row cannot overlap. An axis (or sub-axis) cannot be used to partition one tensor many times. If an axis is explicitly replicated, we cannot use it to partition the tensor.
+
+Thus, each cell represents a factor sharding. A factor can be missing in partial tensors. The table for `C = dot(A, B)` is below. The cells containing an `X` imply that the factor is not in the tensor. For example, F2 is in T1 and T2, but not in T0.
```

Comment:
Could we replace "X" with other symbol to represent that the entry is n/a in the table? We may use sharding axis "x" in the table. Using "X" may be confusing.

Reply:
Good point, I'll think of something.

Reply:
Changed it to N as it may be even better for clarity. Let me know if you have other suggestions.